packages/jupyter-chat/src/model.ts

```diff
@@ -166,6 +172,26 @@ export interface IChatModel extends IDisposable {
    */
   focusInput(): void;
 
+  /**
+   * Add attachment to the next message to send.
+   */
+  addAttachment?(attachment: IAttachment): void;
+
+  /**
+   * Remove attachment to the next message to send.
+   */
+  removeAttachment?(attachment: IAttachment): void;
+
+  /**
+   * Update attachments.
+   */
+  updateAttachments?(attachments: IAttachment[]): void;
+
+  /**
+   * Open attachments.
+   */
+  clickAttachment?(attachments: IAttachment): void;
```

Comment:
Is this the responsibility of the chat model? Opening an attachment doesn't affect chat state, so IMO this method doesn't need to be defined in `IChatModel`. 

Can you explore modifying `AttachmentsComponent` to handle opening attachments directly through the command registry? Then this method and its implementation in `ChatModel` can be removed.

Reply:
Agree, in addition a consumers of the chat widget or the jupyterlab-chat extension should be able to handle other attachment types to other actions / commands.

An option could be to add a registry to the chat widget, linking the function to be called according to the type of attachment (like a `Map<string, (attachment: IAttachment) => void>`).
There could be a token to expose this registry from `jupyterlab-chat`, to allow external extensions (e.g. `jupyter-ai`) to register new behavior with new attachment type (like `var`).

If the registry is not provided with the chat widget options, it could lead to a default one that can open files (as soon as a command registry is provided, at least).

Reply:
Would it be helpful to build some kind of `JupyterlabContext` React context which provides access to all the APIs provided by other JupyterLab extensions?

For example, I was thinking that a component could call something like:


```ts
const { commandRegistry } = useJupyterlabContext();
```

and similarly so to access other JupyterLab singletons. This would simplify the frontend code by not requiring us to pass every singleton as a prop to every child component. This is also equivalent in terms of performance, as these singleton objects are not changed at runtime.

Reply:
@dlqqq I added a registry to add opener function for a given attachment type.
There is a default opener for  `file` type provided by `jupyterlab-chat`, which is available with a token to register more types in external extensions.

As you suggested, the registry is passed back to the react components using a context.
Indeed, we should extend the context to other singletons, probably all the props in [`IChatBodyProps`](https://github.com/jupyterlab/jupyter-chat/blob/73512c16a6c602b81f873ecf9a3c20cb73cefd94/packages/jupyter-chat/src/components/chat.tsx#L111)